src/cdomains/addressDomain.ml

```diff
@@ -20,12 +20,19 @@ end
 
 module AddressSet (Idx: IntDomain.Z) =
 struct
+  module BaseAddr = Lval.BaseAddrRepr (Idx)
   module Addr = Lval.NormalLatRepr (Idx)
-  module J = SetDomain.Joined (Addr)
+  module J = (struct
+    include SetDomain.Joined (Addr)
+    let may_be_equal a b = Option.value (Addr.semantic_equal a b) ~default:true
+  end)
+  module OffsetSplit = DisjointDomain.ProjectiveSetPairwiseMeet (Addr) (J) (Addr.R)
+
   (* module H = HoareDomain.SetEM (Addr) *)
   (* Hoare set for bucket doesn't play well with StrPtr limiting:
      https://github.com/goblint/analyzer/pull/808 *)
-  include DisjointDomain.ProjectiveSet (Addr) (J) (Addr.R)
+  module AddressSet : SetDomain.S with type elt = Addr.t = DisjointDomain.ProjectiveSet (BaseAddr) (OffsetSplit) (BaseAddr.R)
```

Comment:
I have tested this, and this is actually not an issue:
```
%%% apply_invariant: (tests/regression/66-addresses/01-meet.c:18:5-18:12)   oldv: {?, NULL}, newv: {&((alloc@sid:91@tid:[main](#top)), tests/regression/66-addresses/01-meet.c:7:5-7:30)}, meet: {&((alloc@sid:91@tid:[main](#top)), tests/regression/66-addresses/01-meet.c:7:5-7:30)}
```

The reason why this is not an issue is that the `meet` in the address domain is defined via  a `merge`, which checks whether the operands are `top`. In case exactly one of the operands of the `meet` is `top`, the other one is returned (possibly with the nullpointer removed).

Reply:
Right, I think I forgot that there's extra wrapping of `meet` at the end of `AddressSet`.

Also, I think when I wrote this comment, I was still thinking about the structure using one projective set with the original representatives, not the two nested projective sets.